paper.tex

```diff
@@ -67,6 +67,82 @@ \subsection{Path encoding}
+
+There are generally three approaches to such curve approximation. The most straightforward but also least efficient is ``cut then measure,'' usually combined with adaptive subdivision. In this technique, a candidate approximate curve is produced, then the error is measured against the source curve, usually by sampling a number of points along both curves and determining a maximum error (or perhaps some other error norm). If the error is within tolerance, the approximation is accepted. Otherwise, the curve is subdivided (usually at t = 0.5) and each half is recursively approximated. A substantial fraction of all curve approximation methods in the literature are of this form, including \cite{Nehab2020}. The main disadvantage is the cost of computing the error metric. Another risk is underestimating the error due to inadequate sampling; this is a particular problem when the source curve contains a cusp.
+
+% Note: would be cool to run profiler on Nehab code to determine fraction of time computing error metric.
+
+The next approach is similar, but uses an \emph{error metric} to estimate the error. Ideally such a metric is a closed-form computation rather than requiring iteration. A good error metric is conservative, yet tight, in that it never underestimates the error (which would allow results exceeding the error bound to slip through), and does not significantly overestimate the error, which would result in more subdivision than optimal.
+
+By far the most efficient approach is an \emph{invertible} error metric. In this approach, the error metric has an analytic inverse, or at least a good numerical approximation. Because the metric is invertible, it can predict the number of subdivisions needed, as well as the parameter value for each subdivision. If the error metric is accurate, then approximation is near-optimal. One example of an invertible error metric is angle step, used in polar stroking\cite{Kilgard2020}; the number of subdivisions is the total angle subsumed by the curve divided by the angle step size, and the parameter value for each subdivision is the result of solving for a tangent direction. Another widely used invertible error metric is Wang's formula (\cite{Goldman2003}, Section 5.6.3), which gives a bound on the flattening error based on the second derivative of the curve. This metric is conservative but works well in practice; among other applications, it is used in Skia for path flattening. One limitation of Wang's method is that it only applies to the flattening of filled outlines, not parallel curves as is required for strokes.
```

Comment:
> ...among other applications, it is used in Skia for path flattening. One limitation of Wang's method is that it only applies to the flattening of filled outlines, not parallel curves as is required for strokes.

There is potentially more to be said here. My first thought is about citation, since we're mentioning Skia, should we reference code? (I know where to point to here if necessary.)

My second thought is about the fact that Skia actually uses Wang's formula for tessellated GPU strokes using an algorithm that's very much akin to polar stroking. Graphite/Ganesh splits a given cubic into convex sub-cubics (aka "chopping") by analyzing cusps, 180 turns, and inflection points on the CPU. Each chopped segment gets encoded as a separate instance in an instanced triangle-strip draw (alongside joins and caps, which end up in their own instances). The vertex counts for any given instance is such that **a)** they don't exceed the lower bound returned by Wang's formula and **b)** there are always enough vertices to tessellate a 180&deg; radial span (and then some) within the given tolerances. This use of Wang's formula is suitable for polar stroking in general, as the Kilgard'2020 states about their method:

> The number of tessellated quads must also be determined _a priori_ to tessellation of a given path segment, as opposed to being the result of a recursive process; this is motivated by wanting to map well to GPUs where predictable work creation is necessary as GPUs do not naturally support recursive processes.

I wanted to point this out because this scheme can draw strokes correctly (in the strongly correct sense) without computing a precise parallel curve which our text states as _required_. The sentence is not wrong per se but I think there is some nuance wrt the utility of Wang's formula and stroke rendering.[^1]

[^1]: This method of course has its down sides: the CPU analysis isn't cheap and the instance vertex counts are not far from what we would call "optimal" (many triangles are degenerate and are left unused), though there isn't a strong indication that the latter point (the suboptimal-ness) has a significant impact on performance. I should collect data on this.

Reply:
I've addressed this in the text. I think there are a bunch of different things going on here:

* In a local method, you can totally win by outputting a combination a combination of lines flattened by naive Wang and pie shapes between them; these pies are the same as joins (and it doesn't matter whether you generate them by further flattening to tris, or, for example, a fragment shader)

* You can fix Wang by doing a hybrid of Wang and polar stroking. That appears to be basically what Skia and Rive do. It represents state of the art, but is harder to talk about and benchmark, because it's not published as a citeable paper. A good case can be made that we should measure quantitatively, including the number of triangles generated, or possibly the number of outer contour edges.